File: 0037-sudoku-solver/0037-sudoku-solver.js
/**
 * @param {character[][]} board
 * @return {void} Do not return anything, modify board in-place instead.
 */
var solveSudoku = function(board) {
    const rows = Array.from({ length: 9 }, () => new Set());
    const cols = Array.from({ length: 9 }, () => new Set());
    const boxes = Array.from({ length: 9 }, () => new Set());

    // Function to calculate the box index
    const getBoxIndex = (row, col) => Math.floor(row / 3) * 3 + Math.floor(col / 3);

    // Initialize the sets with existing board values
    for (let i = 0; i < 9; i++) {
        for (let j = 0; j < 9; j++) {
            if (board[i][j] !== ".") {
                const num = board[i][j];
                rows[i].add(num);
                cols[j].add(num);
                boxes[getBoxIndex(i, j)].add(num);
            }
        }
    }

    const backtrack = (row, col) => {
        if (row === 9) return true; // Solved the entire board
        if (col === 9) return backtrack(row + 1, 0); // Move to the next row
        if (board[row][col] !== ".") return backtrack(row, col + 1); // Skip filled cells

        const boxIndex = getBoxIndex(row, col);

        for (let num = 1; num <= 9; num++) {
            const char = num.toString();

            if (!rows[row].has(char) && !cols[col].has(char) && !boxes[boxIndex].has(char)) {
                // Place the number
                board[row][col] = char;
                rows[row].add(char);
                cols[col].add(char);
                boxes[boxIndex].add(char);

                if (backtrack(row, col + 1)) return true;

                // Undo placement (backtrack)
                board[row][col] = ".";
                rows[row].delete(char);
                cols[col].delete(char);
                boxes[boxIndex].delete(char);
            }
        }

        return false; // No valid number can be placed
    };

    backtrack(0, 0);
};

File: 0037-sudoku-solver/0037-sudoku-solver.py
class Solution:
    def solveSudoku(self, board: List[List[str]]) -> None:
        rows = [set() for _ in range(9)]
        cols = [set() for _ in range(9)]
        boxes = [set() for _ in range(9)]

        for r in range(9):
            for c in range(9):
                if board[r][c] != ".":
                    num = board[r][c]
                    rows[r].add(num)
                    cols[c].add(num)
                    boxes[(r // 3) * 3 + (c // 3)].add(num)

        def is_valid(board, row, col, num):
            return (
                num not in rows[row]
                and num not in cols[col]
                and num not in boxes[(row // 3) * 3 + (col // 3)]
            )

        def backtrack():
            for r in range(9):
                for c in range(9):
                    if board[r][c] == ".":
                        for num in map(str, range(1, 10)):
                            if is_valid(board, r, c, num):
                                board[r][c] = num
                                rows[r].add(num)
                                cols[c].add(num)
                                boxes[(r // 3) * 3 + (c // 3)].add(num)

                                if backtrack():
                                    return True
                                board[r][c] = "."
                                if num in rows[r]:
                                    rows[r].remove(num)
                                if num in cols[c]:
                                    cols[c].remove(num)
                                if num in boxes[(r // 3) * 3 + (c // 3)]:
                                   boxes[(r // 3) * 3 + (c // 3)].remove(num)

                        return False
            return True

        backtrack()

    def solveSudoku_1(self, board: List[List[str]]) -> None:

        def isValid(board, row, col, num):
            for i in range(9):
                if board[row][i] == num:
                    return False
                if board[i][col] == num:
                    return False
                if board[3 * (row // 3) + (i // 3)][3 * (col // 3) + (i % 3)] == num:
                    return False
            return True

        def backtrack():

            for i in range(9):
                for j in range(9):
                    if board[i][j] == ".":
                        for num in map(str, range(1, 10)):
                            if isValid(board, i, j, num):
                                board[i][j] = num

                                if backtrack():
                                    return True
                                board[i][j] = "."
                        return False

            return True

        backtrack()
